src/coreclr/src/jit/flowgraph.cpp

```diff
@@ -3563,6 +3563,36 @@ void Compiler::fgInitBlockVarSets()
     fgBBVarSetsInited = true;
 }
 
+static bool blockNeedsGCPoll(BasicBlock* block)
+{
+    bool blockMayNeedGCPoll = false;
+    for (Statement* stmt = block->FirstNonPhiDef(); stmt != nullptr; stmt = stmt->GetNextStmt())
+    {
+        if ((stmt->GetRootNode()->gtFlags & GTF_CALL) != 0)
+        {
+            for (GenTree* tree = stmt->GetTreeList(); tree != nullptr; tree = tree->gtNext)
+            {
+                if (tree->OperGet() == GT_CALL)
+                {
+                    GenTreeCall* call = tree->AsCall();
+                    if (call->IsUnmanaged())
+                    {
+                        if (!call->IsSuppressGCTransition())
+                        {
+                            // If the block contains regular unmanaged call, we can depend on it
```

Comment:
I like this optimization but this seems like something that can be forgotten or lost in subsequent optimizations. Is there a test or assert that can be added somewhere to ensure this optimization doesn't get forgotten or missed during further JIT work?

Reply:
This code runs after all jit optimizations before rationalization and lowering. `Lowering::LowerSwitch` does some flowgraph manipulations but I don't think the two calls that are in the same basic block (one with `SuppressGCTransition` and a regular PInvoke) can end up in different basic blocks after that.

Reply:
It might be good to add a comment (perhaps in the header) to the effect that this must be done after any transformations that would add control flow between calls.